File: src/PopulationManager.js
var Constants = require('Constants');
var CreepMiner = require('CreepMiner');

function PopulationManager(roomController) {
    this.roomController = roomController;

    this.allCreeps = this.roomController.getRoom().find(FIND_MY_CREEPS);
    this.sortedCreeps = {};
    for (var i in this.allCreeps) {
        var creep = this.allCreeps[i];
        this.getSortedCreeps(creep.name.charAt(0)).push(creep);
    }
}

PopulationManager.prototype.getSortedCreeps = function(id) {
    if (this.sortedCreeps[id] == null) {
        this.sortedCreeps[id] = [];
    }
    return this.sortedCreeps[id];
};

PopulationManager.prototype.actAllCreeps = function() {
    for (var i in this.allCreeps) {
        var creep = this.allCreeps[i];
        var creepClass = this.getCreepClass(creep);
        if (creepClass != false) {
            creepClass.step();
        }
    }
};

PopulationManager.prototype.getCreepClass = function(creep) {
    switch (creep.name.charAt(0)) {
        case Constants.CREEP_MINER:
            return new CreepMiner(this.roomController, creep);
    }

    return false;
};

module.exports = PopulationManager;

File: src/main/java/com/nicktoony/scrAI/Managers/PopulationManager.java
package com.nicktoony.scrAI.Managers;

import com.nicktoony.helpers.Lodash;
import com.nicktoony.helpers.LodashCallback1;
import com.nicktoony.helpers.LodashCallback2;
import com.nicktoony.scrAI.Constants;
import com.nicktoony.scrAI.Controllers.RoomController;
import com.nicktoony.scrAI.World.Creeps.CreepWorker;
import com.nicktoony.scrAI.World.Creeps.CreepMiner;
import com.nicktoony.scrAI.World.Creeps.CreepWrapper;
import com.nicktoony.scrAI.World.Tasks.Task;
import com.nicktoony.screeps.Creep;
import com.nicktoony.screeps.Game;
import com.nicktoony.screeps.GlobalVariables;
import com.nicktoony.screeps.Memory;
import org.stjs.javascript.Array;
import org.stjs.javascript.Global;
import org.stjs.javascript.JSCollections;
import org.stjs.javascript.Map;

/**
 * Created by nick on 26/07/15.
 * var stjs = require("stjs");
 * var Constants = require('Constants');
 * var Lodash = require('lodash');
 * var CreepMiner = require('CreepMiner');
 */
public class PopulationManager extends Manager {
    private Array<CreepWrapper> allCreeps;
    private Map<String, Array<CreepWrapper>> sortedCreeps;

    public PopulationManager(RoomController roomController, Map<String, Object> memory) {
        super(roomController, memory);
        this.allCreeps = new Array<CreepWrapper>();

        Array<Creep> foundCreeps = (Array<Creep>) this.roomController.getRoom().find(GlobalVariables.FIND_MY_CREEPS, null);
        this.sortedCreeps = JSCollections.$map();
        Lodash.forIn(foundCreeps, new LodashCallback1<Creep>() {
            @Override
            public boolean invoke(Creep creep) {
                String type = (String) creep.memory.$get("type");
                CreepWrapper creepWrapper = getCreepWrapper(type, creep);
                if (creepWrapper != null) {
                    getSortedCreeps(type).push(creepWrapper);
                    allCreeps.push(creepWrapper);
                }
                return true;
            }
        }, this);
    }

    @Override
    protected void init() {

    }

    @Override
    public void update() {
        Global.console.log("PopulationManager -> Update");

        Lodash.forIn(Memory.creeps, new LodashCallback2<Creep, String>() {
            @Override
            public boolean invoke(Creep variable1, String variable2) {

                if (Game.creeps.$get(variable2) == null) {
                    Memory.creeps.$delete(variable2);
                }

                return true;
            }
        }, this);
    }

    public Array<CreepWrapper> getSortedCreeps(String id) {
        if (this.sortedCreeps.$get(id) == null) {
            this.sortedCreeps.$put(id, new Array<CreepWrapper>());
        }
        return this.sortedCreeps.$get(id);
    }

    private CreepWrapper getCreepWrapper(String id, Creep creep) {
        if (id == Constants.CREEP_MINER_ID) {
            return new CreepMiner(roomController, creep);
        } else if (id == Constants.CREEP_WORKER_ID) {
            CreepWorker creepWorker = new CreepWorker(roomController, creep);

            if (creepWorker.getTaskId() != null) {
                Task task = roomController.getTasksManager().getTask(creepWorker.getTaskId());
                if (task != null) {
                    task.assignCreep(creepWorker);
                }
            }

            return creepWorker;
        }

        return null;
    }

    public Array<CreepWrapper> getAllCreeps() {
        return allCreeps;
    }
}
